game-core/src/system/ally.rs

```diff
@@ -21,62 +21,48 @@ pub struct Grouper;
 
 impl<'s> System<'s> for Grouper {
     type SystemData = (
-        ReadStorage<'s, Ally>,
+        WriteStorage<'s, Ally>,
         WriteStorage<'s, Transform>,
-        WriteStorage<'s, Parent>,
-        WriteStorage<'s, Player>,
+        ReadStorage<'s, Player>,
         Entities<'s>,
     );
 
-    fn run(
-        &mut self,
-        (allies, mut transforms, mut parents, mut players, entities): Self::SystemData,
-    ) {
-        let get_relative_position = |mut num_allies: u32| {
-            num_allies %= 9;
+    fn run(&mut self, (mut allies, mut transforms, players, entities): Self::SystemData) {
+        let speed = 5.0;
+        let follow_distance = 20.0;
+        let max_distance = 50.0;
 
-            let ind = if num_allies == 4 {
-                num_allies + 1 // skip player position
-            } else {
-                num_allies
-            } as i32;
-
-            let row = ind / 3 - 1;
-            let col = ind % 3 - 1;
-
-            (row as f32 * 32.0, col as f32 * 32.0)
-        };
-
-        let (p_transform, allies_count, p_entity) = {
-            let (t, p, e) = (&transforms, &players, &entities)
+        let p_transform = {
+            let (t, _) = (&transforms, &players)
                 .join()
                 .next()
                 .expect("no player found");
-            (t.clone(), p.num_allies, e)
+            (t.clone())
         };
 
-        let mut orphans = vec![];
-
-        for (_, _, transform, entity) in (&allies, !&parents, &mut transforms, &entities).join() {
+        for (ally, transform, _entity) in (&mut allies, &mut transforms, &entities).join() {
             let d = p_transform.translation - transform.translation;
-            let merge_dist = f32::powf(32.0 * 1.0, 2.0);
-            if d.truncate().magnitude2() < merge_dist {
-                let (new_x, new_y) = get_relative_position(allies_count);
-
-                transform.translation.x = new_x;
-                transform.translation.y = new_y;
-
-                orphans.push(entity);
+            let m = d.truncate().magnitude();
+            if ally.has_player {
+                let mut v = d / m; //unit vector
+                                   //less than follow distance, do nothing
+                if m < follow_distance {
+                    v *= 0.0;
+                } else if (m > follow_distance) && (m < max_distance) {
+                    //at or more than the follow distance, play a bit of catchup
+                    v *= (1.0 + (m - follow_distance) / (max_distance - follow_distance)) * speed;
+                } else if m > max_distance {
+                    //max catchup is 2.0
+                    v *= 2.0 * speed;
+                }
+                transform.translation += v;
```

Comment:
From the way I'm reading this, it seems like catchup speed is dependant on framerate. I recommend bringing in the Time system and scaling the distance traveled by the time elapsed.